### PR TITLE
Update quaternion covariance at yaw reset [new]

### PR DESCRIPTION
**Pull request by bresch**
Replaces https://github.com/PX4/PX4-Autopilot/pull/22023

### Solved Problem
Since all the elements of the quaternion are coupled, a yaw reset needs to adjust all the covariances correctly to preserve the existing tilt uncertainty.

### Solution
Extract tilt variance before yaw reset and re-apply it afterwards.

### Changelog Entry
For release notes:
```
Fix covariance yaw reset in EKF2
```

### Test coverage
Unit tests

@Donecle could you please check if this fixes your issue?

### CODE REVIEW

File: src/modules/ekf2/EKF/covariance.cpp [hunk 1]
```diff
@@ -528,14 +528,19 @@ bool Ekf::checkAndFixCovarianceUpdate(const SquareMatrixState &KHP)
...
+	float yaw_var = 0.0f;
 
 	// update the yaw angle variance using the variance of the measurement
 	if (PX4_ISFINITE(yaw_noise)) {
 		// using magnetic heading tuning parameter
-		rot_var_ned(2) = (sq(fmaxf(yaw_noise, 1.0e-2f)));
+		yaw_var = (sq(fmaxf(yaw_noise, 1.0e-2f)));
 	}
 
+	resetQuatCov(Vector3f(tilt_var, tilt_var, yaw_var));
```
Comment (dagar) [inline src/modules/ekf2/EKF/covariance.cpp:539]: Should we provide a lower bound here? If the caller doesn't provide a yaw_noise then we're calling resetQuatCov with 0 yaw variance.

File: src/modules/ekf2/EKF/covariance.cpp [hunk 2]
```diff
@@ -528,14 +528,19 @@ bool Ekf::checkAndFixCovarianceUpdate(const SquareMatrixState &KHP)
 void Ekf::resetQuatCov(const float yaw_noise)
 {
 	const float tilt_var = sq(fmaxf(_params.initial_tilt_err, 1.0e-2f));
-	Vector3f rot_var_ned(tilt_var, tilt_var, 0.f);
+	float yaw_var = 0.0f;
```
Comment (dagar) [inline src/modules/ekf2/EKF/covariance.cpp:531]:
```suggestion
	float yaw_var = sq(0.01f);
```

File: src/modules/ekf2/EKF/covariance.cpp [hunk 3]
```diff
@@ -528,14 +528,19 @@ bool Ekf::checkAndFixCovarianceUpdate(const SquareMatrixState &KHP)
...
 
 	// update the yaw angle variance using the variance of the measurement
 	if (PX4_ISFINITE(yaw_noise)) {
 		// using magnetic heading tuning parameter
-		rot_var_ned(2) = (sq(fmaxf(yaw_noise, 1.0e-2f)));
+		yaw_var = (sq(fmaxf(yaw_noise, 1.0e-2f)));
```
Comment (dagar) [inline src/modules/ekf2/EKF/covariance.cpp:536]:
```suggestion
		yaw_var = math::max(sq(yaw_noise), yaw_var);
```

File: src/modules/ekf2/EKF/covariance.cpp (hunk 2 above)
Comment (dagar) [reply src/modules/ekf2/EKF/covariance.cpp:531]: Init to a minimum then if there's a caller provided noise take it below if larger?

File: src/modules/ekf2/EKF/covariance.cpp (hunk 1 above)
Comment (Donecle) [reply src/modules/ekf2/EKF/covariance.cpp:539]: Yes definitely.

**Review by Donecle — APPROVED** [review]
Works fine on my side.

File: src/modules/ekf2/EKF/covariance.cpp (hunk 1 above)
Comment (bresch) [reply src/modules/ekf2/EKF/covariance.cpp:539]: Done

**Review by dagar — APPROVED** [review]

**Pull request merged**